### ISSUE TITLE
Route53: CNAME before R53_ALIAS, invalid batch change

### ISSUE THREAD

**Issue by misilot** [StackExchange/dnscontrol#937]
When trying to add a CNAME and R53_ALIAS in the same update I get the following error

```
CREATE R53_ALIAS dev-system.example.net dev-system18.example.net. atype=CNAME zone_id=<zoneid> ttl=300 type=CNAME zone_id=<zoneid>
CREATE CNAME dev-system18.example ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300

FAILURE! InvalidChangeBatch: [Tried to create an alias that targets dev-system18.example.net., type CNAME in zone <zoneid>, but that target was not found]
	status code: 400, request id: d1a67735-6308-4ab9-abe6-6faf3ce2ac6b
```

However, if I add the CNAME first, and then add the R53_ALIAS it runs fine. Is there a way to prioritize CNAME's before R53_ALIAS in the batches? I am not sure if they would need to be individual batches as well.

Thanks!

**Comment by tlimoncelli** [comment]
I'm not clear about your example.  Was `dev-system18.example` supposed to be `dev-beach18.ksulib.net `?

Could you include the lines from dnsconfig.js? (redacted, of course).

**Comment by misilot** [comment]
Yes, sorry I forgot to update all the references, and sorry I forgot to include the following.

```
    CNAME('dev-system18', 'ec2-54-91-33-155.compute-1.amazonaws.com.'),
    R53_ALIAS('dev-system', 'CNAME', 'dev-system18.ksulib.net.'), 
```

**Comment by tlimoncelli** [comment]
Ah!  I can reproduce this. See https://github.com/StackExchange/dnscontrol/pull/938

**Comment by tlimoncelli** [comment]
Wow, this code is waaaay complicated.

I think the problem is that when an update includes a `R53_ALIAS` the other records in that rrset are removed.  Thus, the update includes just the R53_ALIAS, not the R53_ALIAS and the CNAME.

However the tests are flakey. Sometimes they pass, sometimes they fail.

Here's a sample run:

```
tlimwork4:~/git/dnscontrol/integrationTest$ go test -v -verbose -provider ROUTE53 -run TestDNSProviders -start 32
=== RUN   TestDNSProviders
=== RUN   TestDNSProviders/dnscontroltest-r53.com
=== RUN   TestDNSProviders/dnscontroltest-r53.com/Clean_Slate:Empty
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_cname
    integration_test.go:180: 
        CREATE CNAME dev-system18.dnscontroltest-r53.com ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300
        
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:add_an_alias_to_it
    integration_test.go:180: 
        CREATE R53_ALIAS dev-system.dnscontroltest-r53.com dev-system18.dnscontroltest-r53.com. atype=CNAME zone_id=Z3HTIAEKH5JU4F ttl=300 type=CNAME zone_id=Z3HTIAEKH5JU4F
        
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:Empty
    integration_test.go:180: 
        DELETE CNAME dev-system18.dnscontroltest-r53.com ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300
        DELETE R53_ALIAS dev-system.dnscontroltest-r53.com dev-system18.dnscontroltest-r53.com. atype=CNAME zone_id=Z3HTIAEKH5JU4F ttl=300 type=CNAME zone_id=Z3HTIAEKH5JU4F
        
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_cname+alias_in_one_step
    integration_test.go:180: 
        CREATE CNAME dev-system18.dnscontroltest-r53.com ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300
        CREATE R53_ALIAS dev-system.dnscontroltest-r53.com dev-system18.dnscontroltest-r53.com. atype=CNAME zone_id=Z3HTIAEKH5JU4F ttl=300 type=CNAME zone_id=Z3HTIAEKH5JU4F
        
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:Empty#01
    integration_test.go:180: 
        DELETE R53_ALIAS dev-system.dnscontroltest-r53.com dev-system18.dnscontroltest-r53.com. atype=CNAME zone_id=Z3HTIAEKH5JU4F ttl=300 type=CNAME zone_id=Z3HTIAEKH5JU4F
        DELETE CNAME dev-system18.dnscontroltest-r53.com ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300
        
=== RUN   TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_alias+cname_in_one_step
    integration_test.go:180: 
        CREATE R53_ALIAS dev-system.dnscontroltest-r53.com dev-system18.dnscontroltest-r53.com atype=CNAME zone_id=Z3HTIAEKH5JU4F ttl=300 type=CNAME zone_id=Z3HTIAEKH5JU4F
        CREATE CNAME dev-system18.dnscontroltest-r53.com ec2-54-91-33-155.compute-1.amazonaws.com. ttl=300
        
    integration_test.go:184: InvalidChangeBatch: [Tried to create an alias that targets dev-system18.dnscontroltest-r53.com., type CNAME in zone Z3HTIAEKH5JU4F, but that target was not found]
        	status code: 400, request id: 151bc615-a8ab-4c7d-bf03-be26f9ee02c8
=== RUN   TestDNSProviders/dnscontroltest-r53.com/Post_cleanup:Empty
--- FAIL: TestDNSProviders (2.83s)
    --- FAIL: TestDNSProviders/dnscontroltest-r53.com (2.50s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/Clean_Slate:Empty (0.12s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_cname (0.35s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:add_an_alias_to_it (0.40s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:Empty (0.35s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_cname+alias_in_one_step (0.49s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:Empty#01 (0.28s)
        --- FAIL: TestDNSProviders/dnscontroltest-r53.com/32:R53_ALIAS_ORDER:create_alias+cname_in_one_step (0.20s)
        --- PASS: TestDNSProviders/dnscontroltest-r53.com/Post_cleanup:Empty (0.19s)
FAIL
exit status 1
FAIL	github.com/StackExchange/dnscontrol/v3/integrationTest	3.116s
tlimwork4:~/git/dnscontrol/integrationTest$ 
```

I'm CC:ing the people that have touched the code recently.  

CC: @masterzen @maclennann @rileyberton

**Comment by tlimoncelli** [comment]
That was one heck of a bug!

**Comment by masterzen** [comment]
@tlimoncelli, 

I'm very sorry, I didn't had the time to have a proper look, and you were faster than me :) 

On a side note, my use case with `R53_ALIAS` is to always point to AWS resources (which obviously always pre-exist). I didn't even thought it was possible to point to a `CNAME`.
Thanks for the fix.

**Comment by misilot** [comment]
Thank you @tlimoncelli!

**Comment by tlimoncelli** [comment]
No need to apologize... it gave me an excuse to write [a blog post about it](https://www.yesthatblog.com/post/0085-refactoring-a-loop-that-was-trying-too-hard/).

Once I was able to repro the bug, the issue was such a puzzle I couldn't let it go.

It turns out the code had gotten very complex over the years and was becoming difficult to debug or understand.  The key insight was when I realized:

```go
for range {
    if invariant {
        X ; Xran = true
    } else {
        Y ; Yran = true
    }
}
if Xran { Xfollowup }
if Yran { Yfollowup }
```

should be refactored so that the `if` was on the outside of the loop:

```go
if invariant {
    for range {
        X
    }
    Xfollowup
} else {
    for range {
        Y
    }
    Yfollowup
}
```

Now we could move the Yfollowup code into the for loop, which was the fundamental problem.
